Handle NotImplemented error in service_controller.

prameshj opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds support in service_controller to handle NotImplemented errors when creating a LoadBalancer. CloudProviders using alternate controllers to implement LoadBalancer services or choosing to skip creation of LoadBalancer based on service annotations can now return a NotImplemented error. service_controller will ignore the error without generating an event and will also not modify the Service status in any way.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

prameshj committed:
Handle ImplementedElsewhere error in service_controller

This is used when the cloudprovider layer is not implementing loadBalancer service.
Implementation will be in a different controller running on master.

prameshj committed:
skip ilb creation if subsetting is enabled.

prameshj commented:
/assign @bowei

prameshj commented:
cc @freehan @MrHohn

freehan reviewed (approved):
/lgtm

prameshj commented:
/retest

cheftako commented on pkg/controller/service/service_controller.go:
```diff
@@ -339,6 +339,10 @@ func (s *ServiceController) syncLoadBalancerIfNeeded(service *v1.Service, key st
 		}
 		newStatus, err = s.ensureLoadBalancer(service)
 		if err != nil {
+			if err == cloudprovider.NotImplemented {
```
The NotImplemented error is an already existing error which cloud providers outside of tree may be using. I do not think we should ignore it here. If we want an error to indicate the functionality is implemented  elsewhere lets generate a new error for that case.

prameshj replied:
ack, will add a new error.

prameshj committed:
Add a new error type in cloud.go

Adding a new error "ImplementedElsewhere" to cloud.go.
This error indicates that implementation for a particular service/loadbalancer spec is handled by a different controller.
The caller can ignore the error and skip modifying service status upon receiving this error.

prameshj committed:
unit test

bowei commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -199,9 +199,10 @@ type Routes interface {
-	InstanceNotFound = errors.New("instance not found")
-	DiskNotFound     = errors.New("disk is not found")
-	NotImplemented   = errors.New("unimplemented")
+	InstanceNotFound     = errors.New("instance not found")
```
sort lines

prameshj replied:
Done

bowei commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -199,9 +199,10 @@ type Routes interface {
 var (
-	InstanceNotFound = errors.New("instance not found")
-	DiskNotFound     = errors.New("disk is not found")
-	NotImplemented   = errors.New("unimplemented")
```
unimplemented => not implemented

prameshj replied:
Is there value in changing text of an existing error? There is some code like 

https://github.com/kubernetes/kubernetes/blob/269183a7a83aa48d3a1da43d57e3f9e94e459bd6/staging/src/k8s.io/legacy-cloud-providers/azure/azure_fakes.go#L874

using this string value.

bowei replied:
ah -- yuck. Ok, it's unfortunate that there is no consistency, but let's keep it the same.

bowei commented on pkg/controller/service/service_controller.go:
```diff
@@ -703,7 +707,12 @@ func (s *ServiceController) lockedUpdateLoadBalancerHosts(service *v1.Service, h
 	}
-
+	if err == cloudprovider.ImplementedElsewhere {
+		// Skip error since LoadBalancer implementation is in some other controller. In this case, the loadBalancer will likely not
```
comment is kind of confusing.

```
ImplementedElsewhere indicates that the UpdateLoadBalancer is a nop and the
functionality is implemented by a different controller.  In this case, we
return immediately without doing anything.
```

prameshj replied:
Done

bowei commented:
/hold

Seems to be touch many more files than expected. Can we make sure the codegen is not messed up somehow?

prameshj committed:
addressed review comments

prameshj commented:
/retest

cheftako commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -199,9 +199,10 @@ type Routes interface {
-	DiskNotFound     = errors.New("disk is not found")
-	NotImplemented   = errors.New("unimplemented")
+	DiskNotFound         = errors.New("disk is not found")
+	ImplementedElsewhere = errors.New("Implemented by a different controller")
```
Please do not capitalize the first letter of an error.

prameshj replied:
done

prameshj committed:
fixed error message

prameshj commented:
/retest

prameshj commented:
/retest

cheftako commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -199,9 +199,10 @@ type Routes interface {
-	DiskNotFound     = errors.New("disk is not found")
-	NotImplemented   = errors.New("unimplemented")
+	DiskNotFound         = errors.New("disk is not found")
+	ImplementedElsewhere = errors.New("implemented by a different controller")
```
nit: Your use case is an alternate controller but I think we might want to be more general. "implemented by alternate to cloud provider"?

cheftako commented:
Please make sure to run ./hack/update-gofmt.sh and fix the warnings indicated by ./hack/verify-golint.sh

bowei commented on pkg/controller/service/service_controller.go:
```diff
@@ -339,6 +339,11 @@ func (s *ServiceController) syncLoadBalancerIfNeeded(service *v1.Service, key st
 		newStatus, err = s.ensureLoadBalancer(service)
 		if err != nil {
+			if err == cloudprovider.ImplementedElsewhere {
+				klog.Infof("LoadBalancer for service %s implemented by a different controller %s, Ignoring error",
```
we should probably not log all the time, or at least log at level V(3)

prameshj committed:
Fixed review comments, lint.

Added error check to EnsureLoadBalancerDeleted.

bowei commented:
Don't we need to do the same for EnsureLoadBalancerDeleted

prameshj commented:
> Don't we need to do the same for EnsureLoadBalancerDeleted

I left the delete path as it is, for migration scenario. If there are any pending resources created by service_controller that aren't deleted by the alternate controller, the EnsureLoadBalancerDeleted path will delete those.

prameshj commented:
> > Don't we need to do the same for EnsureLoadBalancerDeleted
> 
> I left the delete path as it is, for migration scenario. If there are any pending resources created by service_controller that aren't deleted by the alternate controller, the EnsureLoadBalancerDeleted path will delete those.

Added check for error returned by EnsureLoadBalancerDeleted. syncLoadBalancerIfNeeded invokes EnsureLoadBalancerDeleted after checking if the loadBalancer exists. If the implementation is handled elsewhere, likely the GetLoadBalancer call is going to return "NotFound", so the delete codepath will not be triggered. However, if the alternate implementation created the loadbalancer with same name as the cloudprovider, then the delete function will be called and the ImplementedElsewhere error should be returned in this case.

prameshj commented:
Just want to clarify the behavior - There are 5 member functions in the cloud.LoadBalancer interface:

GetLoadBalancer - In case of alternate implementation, assumption is that this function will return NotFound
GetLoadBalancerName - In case of alternate implementation, assumption is that this will return empty string
EnsureLoadBalancer, UpdateLoadBalancer, EnsureLoadBalancerDeleted - All three can return NotImplemented error and this is handled by service_controller.go

cheftako commented on pkg/controller/service/service_controller.go:
```diff
@@ -311,6 +311,11 @@ func (s *ServiceController) syncLoadBalancerIfNeeded(service *v1.Service, key st
 			s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletingLoadBalancer", "Deleting load balancer")
 			if err := s.balancer.EnsureLoadBalancerDeleted(context.TODO(), s.clusterName, service); err != nil {
+				if err == cloudprovider.ImplementedElsewhere {
+					klog.V(3).Infof("LoadBalancer for service %s implemented by a different controller %s," +
```
I think Bowei already commented on this but I'm not seeing the comment. Please lower this to log level 4 or lower.

cheftako commented on pkg/controller/service/service_controller.go:
```diff
@@ -339,6 +344,11 @@ func (s *ServiceController) syncLoadBalancerIfNeeded(service *v1.Service, key st
 		newStatus, err = s.ensureLoadBalancer(service)
 		if err != nil {
+			if err == cloudprovider.ImplementedElsewhere {
+				klog.V(3).Infof("LoadBalancer for service %s implemented by a different controller %s, Ignoring error",
```
again

prameshj committed:
Added comment on how to use ImplementedElsewhere

prameshj committed:
Setting log level to 4 since default is 2 or 3.

Depending on how the cluster is created. Test clusters set a default
level of 4.

bowei commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -104,6 +104,12 @@ func GetInstanceProviderID(ctx context.Context, cloud Interface, nodeName types.
 }
 
 // LoadBalancer is an abstract, pluggable interface for load balancers.
+// PR https://github.com/kubernetes/kubernetes/pull/80660 added support for cloud providers to return an error
```
```
LoadBalancer is an abstract, pluggable interface for load balancers.

Cloud provider may chose to implement the logic for
constructing/destroying specific kinds of load balancers in a
controller separate from the ServiceController.  If this is the case,
then {Ensure,Update}LoadBalancer and EnsureLoadBalancerDeleted must
return the ImplementedElsewhere error.  For the given LB, the
GetLoadBalancer must return NotFound and GetLoadBalancerName() return
empty string. EnsureLoadBalancerDeleted must not return
ImplementedElsewhere to ensure proper teardown of resources that
allocated by the ServiceController. (This can happen if a user changes
the type of LB via an update to the resource.)
```

prameshj committed:
Updated comment about ImplementedElsewhere

Removed handling ImplementedElsewhere error in call to EnsureLoadBalancerDeleted.

prameshj committed:
Moved nil check inside AlphaFeatureGate.Enabled

prameshj commented:
/hold cancel

bowei commented on pkg/controller/service/service_controller.go:
```diff
@@ -339,6 +339,11 @@ func (s *ServiceController) syncLoadBalancerIfNeeded(service *v1.Service, key st
 		newStatus, err = s.ensureLoadBalancer(service)
 		if err != nil {
+			if err == cloudprovider.ImplementedElsewhere {
+				klog.V(4).Infof("LoadBalancer for service %s implemented by a different controller %s, Ignoring error",
```
minor: we usually don't break newline inside statement.

prameshj replied:
done

bowei commented on pkg/controller/service/service_controller.go:
```diff
@@ -703,7 +708,12 @@ func (s *ServiceController) lockedUpdateLoadBalancerHosts(service *v1.Service, h
 	}
-
+	if err == cloudprovider.ImplementedElsewhere {
+		// ImplementedElsewhere indicates that the UpdateLoadBalancer is a nop and the
```
Don't you need the comment above as well.

prameshj replied:
done

prameshj committed:
Added comment after invoking ensureLoadBalancer

bowei commented:
/ok-to-test
/lgtm

prameshj commented:
/priority important-soon

prameshj commented:
/assign @cheftako @bowei  for approval.

bowei commented:
/approve

cheftako commented:
/approve

cheftako commented:
/lgtm
